File: 06_Algorithm/leetcode_offer044.go
/**
 * File              : leetcode_offer044.go
 * Author            : Andy
 * Date              : 2022.10.11
 * Last Modified Date: 2022.10.11
 * Last Modified By  : Andy
 */

package largestValues


func max(arr []int) int{
    maxVal := arr[0]
    for i:=1; i < len(arr); i++{
        if arr[i] > maxVal{
            maxVal = arr[i]
        }
    }
    return maxVal
}

func largestValues(root []*TreeNode) []int {
	if root == nil {
		return
	}

    stack := []*TreeNode{root}
    ans := []int{}

    while len(stack) != 0{
        curLevel := make([]int, 0)
        nextLevel := make([]*TreeNode, 0)

        size := len(stack)

        for size != 0; size --{
            cur := stack[size-1]
            stack := stack[:size-1]
            curLevel = append(cur.Val)
            if cur.Left != nil{
                nextLevel = append(nextLevel, cur.Left)
            }

            if cur.Right != nil{
                nextLevel = append(nextLevel, cur.Right)
            }
        }
        ans = append(ans, max(curLevel))
        stack = nextLevel
    }
    return ans
}

File: 06_Algorithm/leetcode_offer044.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_offer044.py
# Author            : Andy
# Date              : 2022.10.11
# Last Modified Date: 2022.10.11
# Last Modified By  : Andy


# 典型的层序遍历，bfs
class Solution:

    def largestValues(self, root):
        if not root:
            return []

        stack = [root]
        ans = []

        while stack:
            cur_level = []
            next_level = []
            size = len(stack)
            while size:
                cur = stack.pop()
                cur_level.append(cur.val)
                if cur.left:
                    next_level.append(cur.left)
                if cur.right:
                    next_level.append(cur.right)
                size -= 1
            ans.append(max(cur_level))
            stack = next_level
        return ans
